src/SAML2/Binding/AbstractHttpBinding.php

```diff
@@ -0,0 +1,298 @@
+
+
+use AdactiveSas\Saml2BridgeBundle\Exception\BadRequestHttpException;
+use AdactiveSas\Saml2BridgeBundle\Exception\InvalidArgumentException;
+use AdactiveSas\Saml2BridgeBundle\Exception\LogicException;
+use AdactiveSas\Saml2BridgeBundle\SAML2\Binding\Exception\UnsupportedBindingException;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+abstract class AbstractHttpBinding implements HttpBindingInterface
+{
+    /**
+     * Validate the signature.
+     *
+     * Throws an exception if we are unable to validate the signature.
+     *
+     * @param ReceivedMessageQueryString $query g.
+     * @param \XMLSecurityKey $key The key we should validate the query against.
+     * @throws BadRequestHttpException
+     * @throws \AdactiveSas\Saml2BridgeBundle\Exception\LogicException
+     * @throws \AdactiveSas\Saml2BridgeBundle\Exception\RuntimeException
+     */
+    public static function validateSignature(ReceivedMessageQueryString $query, \XMLSecurityKey $key)
+    {
+        $algo = urldecode($query->getSignatureAlgorithm());
+
+        if ($key->getAlgorithm() !== $algo) {
+            $key = \SAML2_Utils::castKey($key, $algo);
+        }
+
+        if (!$key->verifySignature($query->getSignedQueryString(), $query->getDecodedSignature())) {
+            throw new BadRequestHttpException(
+                'The SAMLRequest has been signed, but the signature could not be validated'
+            );
+        }
+    }
+
+
+    /**
+     * @param \SAML2_Request $request
+     * @return Response
+     * @throws \InvalidArgumentException
+     * @throws \AdactiveSas\Saml2BridgeBundle\Exception\LogicException
+     */
+    public function getSignedRequest(\SAML2_Request $request)
+    {
+        $destination = $request->getDestination();
+        if($destination === null){
+            throw new LogicException('Invalid destination');
+        }
+
+        $securityKey = $request->getSignatureKey();
+        if($securityKey === null){
+            throw new LogicException('Signature key is required');
+        }
+
+        $requestAsXml = $request->toUnsignedXML()->ownerDocument->saveXML();
+        $encodedRequest = base64_encode(gzdeflate($requestAsXml));
+
+        /* Build the query string. */
+
+        $msg = 'SAMLRequest=' . urlencode($encodedRequest);
+
+        if ($request->getRelayState() !== NULL) {
+            $msg .= '&RelayState=' . urlencode($request->getRelayState());
+        }
+
+        /* Add the signature. */
+        $msg .= '&SigAlg=' . urlencode($securityKey->type);
+
+        $signature = $securityKey->signData($msg);
+        $msg .= '&Signature=' . urlencode(base64_encode($signature));
+
+        if (strpos($destination, '?') === FALSE) {
+            $destination .= '?' . $msg;
+        } else {
+            $destination .= '&' . $msg;
+        }
+
+        return new RedirectResponse($destination);
```

Comment:
This is not abstract, I think the separation should be:
- HttpRedirectBinding: Handle GET method for receive saml request/response and create saml request/response
- HttpPostBinding: Handle POST method for receive saml request/response and create saml request/response

So `HttpPostBinding::getSignedRequest` should generate POST response (using form auto submited in the client)